### PR TITLE
Attempt accessing only valid item properties

### LINKED ISSUE #6939: Liquid `where` filter should fail gracefully when property does not exist
- [x] I believe this to be a bug, not a question about using Jekyll.
- [x] I updated to the latest Jekyll (or) if on GitHub Pages to the latest `github-pages`
- [x] I ran `jekyll doctor` to check my configuration
- [x] I read the CONTRIBUTION file at https://jekyllrb.com/docs/contributing/
- [x] This is a feature request.

---

- [x] I am on (or have tested on) ***macOS*** 10+
- [ ] I am on (or have tested on) ***Debian/Ubuntu*** GNU/Linux
- [ ] I am on (or have tested on) ***Fedora*** GNU/Linux
- [ ] I am on (or have tested on) ***Arch*** GNU/Linux
- [ ] I am on (or have tested on) ***Other*** GNU/Linux
- [ ] I am on (or have tested on) ***Windows*** 10+

---

- [ ] I was trying to install.
- [ ] There is a broken Plugin API.
- [ ] I had an error on GitHub Pages, and I have reproduced it locally.
- [ ] I had an error on GitHub Pages, and GitHub Support said it was a Jekyll Bug.
- [ ] I had an error on GitHub Pages and I did not test it locally.
- [x] I was trying to build.
- [ ] It was another bug.

## My Reproduction Steps

I have a number of pages that have a `context` value that may look like the following:

```yaml
{"Name"=>"Full Stack Web Development", "Curriculum Sets"=>["recaXooVhMqOWC7vJ", "recxZPsJ4zlmCbY6x", "recvV85PJcBALdsBg", "recneLUBwwQw4xPAs"], "Short ID"=>"cs-fsw", "type"=>"course", "id"=>"recUjPmBMgGVAwSA2", "base"=>"lambda-school-curriculum"}
```

I want to filter `site.pages` by the `context.type`, using the following code:

```
{% assign course_pages = site.pages | where: "context.type", "course" %}
```

Jekyll fails with the following trace:

```
  Liquid Exception: undefined method `[]' for nil:NilClass in /_layouts/default.html
bundler: failed to load command: jekyll (/Users/calebhicks/.rbenv/versions/2.4.1/bin/jekyll)
NoMethodError: undefined method `[]' for nil:NilClass
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/lib/jekyll/filters.rb:374:in `block in item_property'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/lib/jekyll/filters.rb:373:in `each'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/lib/jekyll/filters.rb:373:in `reduce'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/lib/jekyll/filters.rb:373:in `item_property'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/lib/jekyll/filters.rb:226:in `block in where'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/lib/jekyll/filters.rb:225:in `select'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/lib/jekyll/filters.rb:225:in `where'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/liquid-4.0.0/lib/liquid/strainer.rb:56:in `invoke'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/liquid-4.0.0/lib/liquid/context.rb:86:in `invoke'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/liquid-4.0.0/lib/liquid/variable.rb:78:in `block in render'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/liquid-4.0.0/lib/liquid/variable.rb:76:in `each'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/liquid-4.0.0/lib/liquid/variable.rb:76:in `inject'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/liquid-4.0.0/lib/liquid/variable.rb:76:in `render'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/liquid-4.0.0/lib/liquid/tags/assign.rb:24:in `render'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/liquid-4.0.0/lib/liquid/block_body.rb:109:in `render_node'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/liquid-4.0.0/lib/liquid/block_body.rb:88:in `block in render'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/liquid-4.0.0/lib/liquid/block_body.rb:75:in `each'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/liquid-4.0.0/lib/liquid/block_body.rb:75:in `render'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/liquid-4.0.0/lib/liquid/template.rb:208:in `block in render'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/liquid-4.0.0/lib/liquid/template.rb:242:in `with_profiling'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/liquid-4.0.0/lib/liquid/template.rb:207:in `render'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/liquid-4.0.0/lib/liquid/template.rb:220:in `render!'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/lib/jekyll/liquid_renderer/file.rb:30:in `block (2 levels) in render!'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/lib/jekyll/liquid_renderer/file.rb:42:in `measure_bytes'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/lib/jekyll/liquid_renderer/file.rb:29:in `block in render!'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/lib/jekyll/liquid_renderer/file.rb:49:in `measure_time'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/lib/jekyll/liquid_renderer/file.rb:28:in `render!'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/lib/jekyll/tags/include.rb:139:in `block in render'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/liquid-4.0.0/lib/liquid/context.rb:123:in `stack'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/lib/jekyll/tags/include.rb:136:in `render'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/liquid-4.0.0/lib/liquid/block_body.rb:109:in `render_node'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/liquid-4.0.0/lib/liquid/block_body.rb:88:in `block in render'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/liquid-4.0.0/lib/liquid/block_body.rb:75:in `each'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/liquid-4.0.0/lib/liquid/block_body.rb:75:in `render'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/liquid-4.0.0/lib/liquid/template.rb:208:in `block in render'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/liquid-4.0.0/lib/liquid/template.rb:242:in `with_profiling'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/liquid-4.0.0/lib/liquid/template.rb:207:in `render'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/liquid-4.0.0/lib/liquid/template.rb:220:in `render!'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/lib/jekyll/liquid_renderer/file.rb:30:in `block (2 levels) in render!'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/lib/jekyll/liquid_renderer/file.rb:42:in `measure_bytes'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/lib/jekyll/liquid_renderer/file.rb:29:in `block in render!'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/lib/jekyll/liquid_renderer/file.rb:49:in `measure_time'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/lib/jekyll/liquid_renderer/file.rb:28:in `render!'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/lib/jekyll/renderer.rb:123:in `render_liquid'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/lib/jekyll/renderer.rb:188:in `render_layout'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/lib/jekyll/renderer.rb:155:in `place_in_layouts'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/lib/jekyll/renderer.rb:85:in `render_document'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/lib/jekyll/renderer.rb:62:in `run'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/lib/jekyll/site.rb:466:in `block in render_pages'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/lib/jekyll/site.rb:464:in `each'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/lib/jekyll/site.rb:464:in `render_pages'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/lib/jekyll/site.rb:193:in `render'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/lib/jekyll/site.rb:73:in `process'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/lib/jekyll/command.rb:28:in `process_site'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/lib/jekyll/commands/build.rb:65:in `build'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/lib/jekyll/commands/build.rb:36:in `process'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/lib/jekyll/commands/serve.rb:42:in `block (3 levels) in init_with_program'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/lib/jekyll/commands/serve.rb:42:in `each'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/lib/jekyll/commands/serve.rb:42:in `block (2 levels) in init_with_program'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/mercenary-0.3.6/lib/mercenary/command.rb:220:in `block in execute'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/mercenary-0.3.6/lib/mercenary/command.rb:220:in `each'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/mercenary-0.3.6/lib/mercenary/command.rb:220:in `execute'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/mercenary-0.3.6/lib/mercenary/program.rb:42:in `go'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/mercenary-0.3.6/lib/mercenary.rb:19:in `program'
  /Users/calebhicks/.rbenv/versions/2.4.1/lib/ruby/gems/2.4.0/gems/jekyll-3.6.2/exe/jekyll:15:in `<top (required)>'
  /Users/calebhicks/.rbenv/versions/2.4.1/bin/jekyll:22:in `load'
  /Users/calebhicks/.rbenv/versions/2.4.1/bin/jekyll:22:in `<top (required)>'
```

## The Output I Wanted

I would expect the `where` filter to fail gracefully when filtering for properties that may not exist on every item in the array, and skip the items that do not have the property.

If this is a Liquid issue, let me know and I can file it there. Thank you!

### PR DESCRIPTION
**Pull request by ashmaroli**
Resolves #6939

### CODE REVIEW

**Comment by kenman345** [comment]
does this need tests with it?

**Comment by mattr-** [comment]
Looks like we need a test for what happens if something isn't an array, at least.

**Comment by ashmaroli** [comment]
@mattr- Since the time the latest update to `where` filter was merged in, I hv been trying to see if we can support getting a hash `foo`'s sub-property via use of `foo.bar` in the `where` filter..
Will open a PR for that this Sunday and let it close this PR organically..

**Comment by mattr-** [comment]
I'm not ready to introduce the complexity for a `foo.bar` style sub-property query for the where filter right now this late in the release process. We can consider it for a future release. I'd rather see this PR ship first.

File: test/test_filters.rb [hunk 1]
```diff
@@ -849,6 +849,20 @@ def to_liquid
         assert_equal 2, @filter.where(array, "color", nil).length
       end
 
+      should "filter array and string objects appropriately" do
```
Comment (mattr-) [inline test/test_filters.rb]: Based on this test, the functionality feels too complicated to fix the bug. The most straightforward way I know to test this is to write something like:
```ruby
should "not throw exception when an item doesn't have the property" do
  array = [{}, { "color" => nil }, { "color" => "" }, { "color" => "text" }, { "foo" => "bar" }]
  assert_equal [{ "color" => "text" }], @filter.where(hash, "color", "text")
end
```

Also, filtering on sub properties, as in #6939 (`context.type`), isn't valid syntax. This feels like we need a better reproduction case.

Comment (ashmaroli) [reply test/test_filters.rb]: Your test-case passes on `master`. The issue outlined in #6939 is actually due to `#item_property` encountering a `nil` element.

Modifying your test-case to the following (will fail on current `master`:
```ruby
should "not throw exception when an item doesn't have the property" do
  array = [{}, { "color" => "text" }, { "foo" => "bar" }, nil]
  assert_equal [{ "color" => "text" }], @filter.where(array, "color", "text")
end
```
Another case that'll fail on `master`:
```ruby
should "not throw exception when an item doesn't have the property" do
  array = [{}, { "color" => "text" }, [], { "foo" => "bar" }]
  assert_equal [{ "color" => "text" }], @filter.where(array, "color", "text")
end
```

**Pull request closed without merging**